[mqtt-cpp] Upgrade to 10.0.0

iainchesworth opened the pull request:
Updates the mqtt-cpp port to v10.0.0.  

- #### What does your PR fix?  
Adds missing support for optional TLS, websocket, and logging features.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline]
No

- #### Does your PR follow the [maintainer guide]
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
Yes

iainchesworth committed:
Upgrade mqttcpp to 10.0.0

Updated mqttcpp portfile to v10.0.0 and added optional features supporting logging, TLS, and websockets.

iainchesworth committed:
Add output from "vcpkg x-add-version --all"

ras0219 commented on ports/mqtt-cpp/portfile.cmake:
```diff
@@ -1,15 +1,37 @@
     HEAD_REF master
 )
 
+if ("log" IN_LIST FEATURES)
```
This is a great place to use `vcpkg_check_features()`:
```cmake
vcpkg_check_features(
    OUT_FEATURE_OPTIONS FEATURE_OPTIONS
    FEATURES
        log MQTT_USE_LOG
        tls MQTT_USE_TLS
        ws MQTT_USE_WS
)

vcpkg_configure_cmake(
  OPTIONS ${FEATURE_OPTIONS}
)
```

iainchesworth committed:
Switch to using vcpkg_check_feature

Instead of using individual flag checks for optional features, switch to using vcpkg_check_features for the set of flags.

iainchesworth committed:
Fix existing broken interface export

The original interface export (mqtt_cpp_iface) in the upstream repo is broken and doesn't export properly under Windows.  This commit fixes the export so that it is correctly located and cmake's "find_package" works.

iainchesworth committed:
Fix incorrect installation of header files that lost folder structure

JonLiu1993 commented on ports/mqtt-cpp/portfile.cmake:
```diff
@@ -1,24 +1,50 @@
     SOURCE_PATH ${SOURCE_PATH}
     PREFER_NINJA
     OPTIONS
+	${OUT_FEATURE_OPTIONS}
```
There should be ${FEATURE_OPTIONS}

JonLiu1993 commented on ports/mqtt-cpp/portfile.cmake:
```diff
@@ -1,24 +1,50 @@
     PREFER_NINJA
     OPTIONS
+	${OUT_FEATURE_OPTIONS}
+	-DMQTT_USE_STR_CHECK=ON
```
Please indent two spaces backward to keep the format consistent with the options that follow

iainchesworth committed:
Fix errors in portfile per PR comments

iainchesworth committed:
Fix tabs/spaces issue so formatting is correct.

iainchesworth commented:
Added requested changes (formatting).

JonLiu1993 committed:
update version

JonLiu1993 commented:
All features are tested successfully in the folllowing triplet:
- x64-windows
- x64-windows-static
- x64-linux

ras0219-msft commented on ports/mqtt-cpp/portfile.cmake:
```diff
@@ -1,24 +1,50 @@
     HEAD_REF master
 )
 
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt DESTINATION ${SOURCE_PATH})
```
By replacing the entire CMake build, this also removes any transitive dependencies needed to consume the library (openssl? boost?). It'd be much better if we could patch the original build to fix the problem instead of replacing it entirely.

JonLiu1993 commented:
@iainchesworth ,Have you considered @ras0219-msft 's suggestion?

PhoebeHui commented:
Ping @iainchesworth for response. Is work still being done for this PR?
